dmscripts/compare_dmp_and_dnb_data.py

```diff
@@ -0,0 +1,339 @@
+                    return
+            else:
+                print(row)  # TODO: log maybe?
+
+    def fetch_and_compare_frameworks_suppliers_data(
+        self,
+        frameworks=['g-cloud', 'digital-outcomes-and-specialists'],
+        csv_filename=None,
+        export_dnb_data=True,
+        export_dmp_data=False,
+        from_declaration=True
+    ):
+        "retrieve DMp suppliers of frameworks and call for each the D&B API to compare and optionally append to CSV"
+        counter = 0
+        # fetch all DUNS numbers from CSV and if it doesn't exist create it
+        if csv_filename is not None:
+            try:
+                with open(csv_filename, 'rb') as f:
+                    reader = unicodecsv.reader(f, encoding='utf-8')
+                    next(reader)  # skip the CSV's header
+                    for row in reader:
+                        print(row)
+                        self.duns_number_compared.add(row[0])
+                    print(self.duns_number_compared)
+            except FileNotFoundError:
+                header = [
+                    'DUNS number',
+                    'Compared at',
+                    'In D&B?',
+                    'Name',
+                    'Company House',
+                    'Country',
+                    'Postcode',
+                    'Address',
+                    'D&B Error',
+                ]
+                if export_dnb_data:
+                    header += [
+                        'D&B Name',
+                        'D&B Company House',
+                        'D&B Country',
+                        'D&B Postcode',
+                        'D&B Address',
+                    ]
+                if export_dmp_data:
+                    header += [
+                        'DMp Name',
+                        'DMp Company House',
+                        'DMp Country',
+                        'DMp Postcode',
+                        'DMp Address',
+                    ]
+                with open(csv_filename, 'wb') as f:
+                    writer = unicodecsv.writer(f, encoding='utf-8')
+                    writer.writerow(header)
+            # open for write later
+            f = open(csv_filename, 'ab')  # file handle is closed at the end of the method
+            writer = unicodecsv.writer(f, encoding='utf-8')
+        # iterate all the suppliers of the frameworks
+        for framework in frameworks:
+            dmp_response = self.get_dmp_supplier_data(framework=framework, from_declaration=from_declaration)
+            if from_declaration:
+                dmp_response = dmp_response.get('supplierFrameworks', [])
+            else:
+                dmp_response = dmp_response.get('suppliers', [])
+            for dmp_supplier in dmp_response:
+                self.compare_suppliers(
+                    dmp_supplier, from_declaration, export_dnb_data, export_dmp_data, counter, writer
+                )
```

Comment:
@risicle I pulled part of the function into a new one, because flake was complaining. Works the same as before. Can you have a look?

Reply:
Ugh.. I expect you're referring to the "complexity checks". I'm very sceptical about them. TBH I think the result it's guided you towards is harder to understand - a _many_-arg'd function which adds the uncertainty over which conditions it's going to be called in. An inner block of a larger function has no such uncertainly - execution can only reach that point by following the function's prologue.

:man_shrugging: 

It is permitted to bump the check's limit, yknow.